### DISCUSSION TITLE
Add Variables To QR Links

### DISCUSSION THREAD

**Discussion by omicreativedev** [deso-protocol/dips#81] [Medium CIPs]
Right now QR codes resolve to the following style links

`https://bitclout.com/send-bitclout?public_key=BC1YLhTk3qzocZRDAAHyFvy7PME3xvPZyJn3eY2tfjGa8GmNwheU8Xv`

This is awesome because you can use links across the web. For instance, on my blog I can use:

`<a href="https://bitclout.com/send-bitclout?public_key=BC1YLhTk3qzocZRDAAHyFvy7PME3xvPZyJn3eY2tfjGa8GmNwheU8Xv">Buy Me A Coffee</a>`

What I propose is we add:

?usd=
or/and
?clout=

So that we can prepopulate the field in the wallet with the amount.

This will allow creators and people make buy, donation, etc. links on their website or blog.

`<a href="https://bitclout.com/send-bitclout?usd=3.50?public_key=BC1YLhTk3qzocZRDAAHyFvy7PME3xvPZyJn3eY2tfjGa8GmNwheU8Xv">Buy Me A Coffee</a>`

We could also introduce a feature on posts where a user can set a CTA (call to action button) embed. The ability to add a simple PAY button of a certain amount that automatically goes to the wallet using the same link where the fields are prepopulated with the amount.

Likewise, extensions (or main) could add a Tip button directly on profiles.

This could also go in hand with my suggestion of adding a 🚀 option next to diamonds.
As suggested here: https://github.com/bitclout/cips/discussions/26

Note: I might need help if this gets an approval.

**Comment by tijno** [comment]
yes!

it would also be good to include this message when sending creator coins or bitclout to someone.

may also relate to https://github.com/bitclout/cips/discussions/46 about custom message when buying/selling a coin

**Reply by omicreativedev** [reply]
I didn't even think of that but you're right. Now I wonder what Nordian and Aikon would think about this in a grander scope type of way.

**Reply by AikonMishima** [reply]
I support this proposal.

It makes sense to have the ability to choose, especially with an elegant parameter change in the url.

As for the message, this was indeed one of the use cases I was thinking about.

in the future, we might not even be buying a Creator Coin from a BitClout.com clone.
So it would make sense to have a payload carrier for the message, baked into the code.

E.g. if I click on your “buy me a coffee” on BrixMusic.com, and I write you a message there, you’ll see the purchase on BitClout.com together with my message.

**Reply by bloated-devish** [reply]
@AikonMishima if you go down this path - I think the ```ExtraData``` is probably the way to go. Can we all agree on the length of "message" as 280 chars (same as length of post - without hacks).

So there are 3 sets of changes:
1) Frontend - URL params sanitization for comments, validation for amount etc. Also for the message, please mention how you will read the creator coin / creator coin transfer notes.
2) Backend - https://github.com/bitclout/backend/blob/main/routes/transaction.go#L1462-L1592 `txn` has ExtraData, you can add this there (this is not super clean, but prevents a fork). Also for all the read paths, you'll have to "reconstruct" message from extradata if one is available.
3) Core - not sure what changes, since we are only stuffing this to ExtraData (which is at tnx level).

-Bloated

PS: @teambrixmusic  - it should be `<a href="https://bitclout.com/send-bitclout?usd=3.50`**&**`public_key=BC1YLhTk3qzocZRDAAHyFvy7PME3xvPZyJn3eY2tfjGa8GmNwheU8Xv&notes='Supporting you to get a coffee'">Buy Me A Coffee</a>`

**Reply by omicreativedev** [reply]
> @AikonMishima if you go down this path - I think the `ExtraData` is probably the way to go. Can we all agree on the length of "message" as 280 chars (same as length of post - without hacks).
> 
> So there are 3 sets of changes:
> 
> 1. Frontend - URL params sanitization for comments, validation for amount etc. Also for the message, please mention how you will read the creator coin / creator coin transfer notes.
> 2. Backend - https://github.com/bitclout/backend/blob/main/routes/transaction.go#L1462-L1592 `txn` has ExtraData, you can add this there (this is not super clean, but prevents a fork). Also for all the read paths, you'll have to "reconstruct" message from extradata if one is available.
> 3. Core - not sure what changes, since we are only stuffing this to ExtraData (which is at tnx level).
> 
> -Bloated
> 
> PS: @teambrixmusic - it should be `<a href="https://bitclout.com/send-bitclout?usd=3.50`**&**`public_key=BC1YLhTk3qzocZRDAAHyFvy7PME3xvPZyJn3eY2tfjGa8GmNwheU8Xv&notes='Supporting you to get a coffee'">Buy Me A Coffee</a>`

255 characters makes sense. Probably could be less than that.  Post length is going to vary based on what the node or dapp is about I'm sure. There will be a lot of different uses. But I think messages being 255 per message is more than sensible.

^ this 2 lines is ~245
